### PR TITLE
Conversion Utility Optimization

### PR DESCRIPTION
**Pull request by snehalv2002**
Speed up the conversion util

# Description

Start with a short description of what the PR does and how this is a change from
the past.

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/123456
FIXES: #123456

You can also provide a comma-separated list. If you don't want to close a bug but
simply to reference it, use BUGS, e.g.: 
BUGS: b/123456

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [ ] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [ ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have run end-to-end tests tests and provided workload links above if applicable.
- [ ] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/checkpoint_conversion/to_maxtext.py [hunk 1]
```diff
@@ -389,29 +496,36 @@ def _build_single_axis_stacked_tensor(
...
+      mt_param_name: The name of the MaxText parameter.
+      enable_parallel_conversion: Whether to parallelize layer transformations across threads.
 
   Returns:
       The final, assembled NumPy array for the MaxText parameter.
   """
   tensors_to_stack = []
 
-  if config.scan_layers:
+  if config.scan_layers and "scanned_blocks" in mt_param_name:
```
Comment (AusarYao28) [inline src/maxtext/checkpoint_conversion/to_maxtext.py:524]: MaxText parameter paths for scanned layers do not always contain "scanned_blocks", probably change to "or"?

File: src/maxtext/checkpoint_conversion/to_huggingface.py [hunk 2]
```diff
@@ -483,7 +525,12 @@ def main(argv: Sequence[str]) -> None:
   if not config.base_output_directory:
     output_directory = f"tmp/{config.run_name}"
   else:
-    output_directory = config.base_output_directory
+    output_directory = os.path.expanduser(config.base_output_directory)
+    if output_directory.startswith("/usr/local/google/home/"):
```
Comment (AusarYao28) [inline src/maxtext/checkpoint_conversion/to_huggingface.py:529]: Hardcoding /usr/local/google/home/ workstation string manipulation into open-source MaxText is an anti-pattern.

File: src/maxtext/checkpoint_conversion/to_maxtext.py [hunk 3]
```diff
@@ -87,6 +90,69 @@
 absl.logging.set_verbosity(absl.logging.INFO)  # for max_logging.log
 
 
+def get_key_mapper(model_name):
```
Comment (AusarYao28) [inline src/maxtext/checkpoint_conversion/to_maxtext.py:93]: I think this mapping belongs in param_mapping.py alongside all other model definitions.

File: src/maxtext/checkpoint_conversion/to_huggingface.py [hunk 4]
```diff
@@ -516,6 +563,8 @@ def main(argv: Sequence[str]) -> None:
 
   # 4. Extract and transform weights for Linen/NNX-SFT/NNX-RL checkpoints
   maxtext_state_dict = detect_and_extract_checkpoint(checkpoint_dict)
+  del checkpoint_dict
+  import gc; gc.collect()
```
Comment (AusarYao28) [inline src/maxtext/checkpoint_conversion/to_huggingface.py:567]: Repeated inline imports should be moved to top-level module imports.

File: src/maxtext/checkpoint_conversion/to_maxtext.py [hunk 5]
```diff
@@ -30,26 +30,29 @@
...
-    scan_layers=True
+    src/maxtext/configs/base.yml model_name="qwen3.5-35b-a3b" \
+    --hf_model_path="/path/to/HF_weights" \
+    base_output_directory="/path/to/output_checkpoint" \
+    scan_layers=True attention=dot_product \
+    --lazy_load_tensors=False --enable_parallel_conversion=True
 """
 
 import argparse
+from concurrent.futures import ThreadPoolExecutor
```
Comment (AusarYao28) [inline src/maxtext/checkpoint_conversion/to_maxtext.py:55]: Repeated inline imports should be moved to top-level module imports.

File: src/maxtext/checkpoint_conversion/to_huggingface.py [hunk 6]
```diff
@@ -396,7 +407,30 @@ def _transform_weights_to_full_model(config, filtered_map_keys, state_dict, para
...
+
+  if not enable_parallel_conversion:
+    processed_params_list = []
+    for key in MemoryMonitorTqdm(filtered_map_keys, leave=True):
+      processed_params_list.extend(_process_single_key(key))
+  else:
+    max_logging.log("Parallel to_huggingface weight transformation ENABLED (Strategies 1, 2, & 3)")
+    num_workers = min(os.cpu_count() or 16, 32)
+    from concurrent.futures import ThreadPoolExecutor
+    with ThreadPoolExecutor(max_workers=num_workers) as executor:
```
Comment (AusarYao28) [inline src/maxtext/checkpoint_conversion/to_huggingface.py:423]: Outer ThreadPoolExecutor iterates over filtered_map_keys. Concurrently, inside each task, process_maxtext_param creates an inner ThreadPoolExecutor over slices. When max_workers == 32, up to 1,024 threads compete for the GIL, while 32 parameter layers are transformed and held in host RAM simultaneously, creating a severe OOM risk on large models.
